platforms/documentation/docs/src/docs/userguide/dep-man/declaring-depedencies/declaring_configurations.adoc

```diff
@@ -0,0 +1,148 @@
+
+====
+include::sample[dir="snippets/dependencyManagement/definingUsingConfigurations-custom/kotlin",files="build.gradle.kts[tags=custom-configuration]"]
+include::sample[dir="snippets/dependencyManagement/definingUsingConfigurations-custom/groovy",files="build.gradle[tags=custom-configuration]"]
+====
+
+You can manage custom configurations using the `configurations` object.
+Configurations can have names and can extend each other.
+For more details, refer to the link:{groovyDslPath}/org.gradle.api.artifacts.ConfigurationContainer.html[`ConfigurationContainer`] API.
+
+Remember that configurations are intended to be used for a single role: declaring dependencies, performing resolution, or defining consumable variants.
+
+[[sec:resolvable-consumable-configs]]
+== Resolvable and consumable configurations
+
+Configurations aren't used just for declaring dependencies, they serve various roles in dependency management:
+
+- *Declaring Dependencies*: Configurations define a set of dependencies.
+- *Consumer Role*: Configurations are used to resolve dependencies into artifacts.
+- *Producer Role*: Configurations expose artifacts for consumption by other projects.
+
+=== Configurations for declaring dependencies
+
+To express that an application `app` _depends on_ library `lib`, _at least_ one configuration is required:
+
+====
+include::sample[dir="snippets/dependencyManagement/attributeMatching/kotlin",files="build.gradle.kts[tags=declare-configuration]"]
+include::sample[dir="snippets/dependencyManagement/attributeMatching/groovy",files="build.gradle[tags=declare-configuration]"]
+====
+
+*Configurations used for declaring dependencies* define and organize the dependencies required for different aspects of a project.
+
+=== Configurations for consumers (resolvable configuration)
+
+To specify how dependencies are used, you can create companion configurations:
+
+====
+include::sample[dir="snippets/dependencyManagement/attributeMatching/kotlin",files="build.gradle.kts[tags=concrete-classpath]"]
+include::sample[dir="snippets/dependencyManagement/attributeMatching/groovy",files="build.gradle[tags=concrete-classpath]"]
+====
+
+Now, `someConfiguration` declares dependencies, while `compileClasspath` and `runtimeClasspath` are resolvable configurations for specific purposes.
+When resolved they should contain the compile classpath, and the runtime classpath of the application respectively.
+
+*Resolvable configurations* are those that can be resolved to produce a set of files or artifacts.
+These configurations are used to define the classpath for different stages of a build process, such as compilation or runtime.
+
+=== Configurations for producers (consumable configuration)
+
+To specify what is exposed to consumers, you can create consumable configurations:
+
+====
+include::sample[dir="snippets/dependencyManagement/attributeMatching/kotlin",files="build.gradle.kts[tags=setup-configurations]"]
```

Comment:
In this example, there are some assertions in the code. I'm not exactly sure why they are there, but since users are not going to write those assertions themselves, we should not include it in the code sample.

Reply:
The assertions were removed, and they were not used by any test.